diff --git a/rxe/rxe_comp.c b/rxe/rxe_comp.c
index 72b30e6..5d036eb 100644
--- a/rxe/rxe_comp.c
+++ b/rxe/rxe_comp.c
@@ -425,7 +425,8 @@ static void make_send_cqe(struct rxe_qp *qp, struct rxe_send_wqe *wqe,
 			uwc->byte_len = wqe->dma.length;
 		}
 	} else {
-		if (wqe->status != IB_WC_WR_FLUSH_ERR)
+		if ((wqe->status != IB_WC_WR_FLUSH_ERR) && 
+			(wqe->status != IB_WC_RETRY_EXC_ERR))
 			rxe_err_qp(qp, "non-flush error status = %d\n",
 				wqe->status);
 	}
